ra2ce/network/avg_speed/avg_speed_calculator.py

```diff
@@ -0,0 +1,215 @@
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+import logging
+from pathlib import Path
+from re import split
+from statistics import mean
+
+import networkx as nx
+
+import ra2ce.network.networks_utils as nut
+from ra2ce.network.avg_speed.avg_speed import AvgSpeed
+from ra2ce.network.avg_speed.avg_speed_reader import AvgSpeedReader
+from ra2ce.network.avg_speed.avg_speed_writer import AvgSpeedWriter
+from ra2ce.network.network_config_data.enums.road_type_enum import RoadTypeEnum
+
+
+class AvgSpeedCalculator:
+    """
+    Class to calculate the average speed of a road network based on the maximum speed of the roads.
+    """
+
+    graph: nx.Graph
+    output_graph_dir: Path | None
+    avg_speed: AvgSpeed
+
+    def __init__(self, graph: nx.Graph, output_graph_dir: Path | None) -> None:
+        self.graph = graph
+        self.output_graph_dir = output_graph_dir
+        self.avg_speed = None
+
+    @staticmethod
+    def parse_speed(speed_input: str | list[str]) -> float:
+        """
+        Parse the average speed from the input string(s).
+
+        Args:
+            speed_input (str | list[str]): (List of) string(s) with the speed(s).
+                Can have different formats, e.g. "50 mph", "50", "50;60", "50-60", "50|60".
+
+        Returns:
+            float: Average speed of the input string(s).
+                0.0 is returned if the input is empty or can't be parsed correctly.
+        """
+        if not speed_input:
+            return 0.0
+        if isinstance(speed_input, list):
+            return mean(map(AvgSpeedCalculator.parse_speed, speed_input))
+        if " mph" in speed_input:
+            return float(speed_input.split(" mph")[0]) * 1.609344
+        if any(c.isalpha() for c in speed_input):
+            return 0.0
+        # Split on the different separators (";" or "-" or "|")
+        # and take the mean of the results (if any).
+        return mean(float(x) for x in split(r";|-|\|", speed_input) if x.isnumeric())
+
+    def _get_avg_speed(
+        self,
+        road_type_col_name: str,
+    ) -> AvgSpeed:
+        def get_main_speed(link_type: RoadTypeEnum) -> float:
+            _base_rt = RoadTypeEnum.get_enum(str(link_type).split("_link")[0])
+            if _base_rt in _avg_speed.road_types:
+                return _avg_speed.get_avg_speed(_base_rt)
+            return 0.0
+
+        def get_uncombined_speed(combined_types: list[RoadTypeEnum]) -> float:
+            for _rt in combined_types:
+                if _rt in _avg_speed.road_types:
+                    return _avg_speed.get_avg_speed(_rt)
+            return 0.0
+
+        _avg_speed = AvgSpeed()
```

Comment:
The idea of internal methods is that they are aware of the context, defining the `_avg_speed` after the methods seems a bit unlogical to me.
```suggestion
        _avg_speed = AvgSpeed()
        def get_main_speed(link_type: RoadTypeEnum) -> float:
            _base_rt = RoadTypeEnum.get_enum(str(link_type).split("_link")[0])
            if _base_rt in _avg_speed.road_types:
                return _avg_speed.get_avg_speed(_base_rt)
            return 0.0

        def get_uncombined_speed(combined_types: list[RoadTypeEnum]) -> float:
            for _rt in combined_types:
                if _rt in _avg_speed.road_types:
                    return _avg_speed.get_avg_speed(_rt)
            return 0.0
```

Reply:
I prefer having the internal methods before the rest of the code. Apparently the context is only to be known at the position of calling.
Let's have a chat.

Reply:
After previous refactoring the internal methods don't use locally defined variables anymore, so they stay on top. Will consider your input for a next case.